activerecord/lib/active_record/connection_adapters/mysql2/database_statements.rb

```diff
@@ -48,26 +48,39 @@ def perform_query(raw_connection, sql, binds, type_casted_binds, prepare:, notif
             # made since we established the connection
             raw_connection.query_options[:database_timezone] = default_timezone
 
-            result = if prepare
-              stmt = @statements[sql] ||= raw_connection.prepare(sql)
+            result = if !prepared_statements || binds.nil? || binds.empty?
+              ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
+                result = raw_connection.query(sql)
+                @affected_rows_before_warnings = raw_connection.affected_rows
+                result
+              end
+              result
+            else
+              if prepare
+                stmt = @statements[sql] ||= raw_connection.prepare(sql)
+              else
+                stmt = raw_connection.prepare(sql)
+              end
 
               begin
                 ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
                   result = stmt.execute(*type_casted_binds)
+                  @affected_rows_before_warnings = stmt.affected_rows
```

Comment:
Using `stmt` here makes sense, but I don't understand how `permit_concurrent_loads` would be relevant

Reply:
I was seeing failures (aaaand the merge commit also flaked), like this one where I added logging on the query:
https://buildkite.com/rails/rails/builds/114048#01935945-5628-45d5-aaa8-ad6535c4069e/1246-1323

That lead me to believe that lock was necessary, and leaning on review to enlighten me.

Reply:
It's an _unlock_. There's definitely something concurrency-ish broken here -- I've seen the same random failures -- but even if this does nudge timings in a way that makes it fail less frequently (which still seems implausible to me, but doesn't matter either way), we'll need to actually understand the nature of the problem.

Even if this were a lock, the sporadic nature of concurrency behaviours means it's critical to properly understand exactly why something needs a lock, not just observe that it seems to make the problem go away.

Reply:
I figured the flaky test, it's a bug in Mysql2: https://github.com/brianmario/mysql2/pull/1383